shared_client.go

```diff
@@ -206,15 +206,27 @@ func handler(wg *sync.WaitGroup, client *Client, conn *Conn, requests chan reque
 		conn.Close()
 		close(receiverTrigger)
 
-		// Drain responses send by receive loop to allow it to exit.
+		// Drain responses sent by receive loop to allow it to exit.
 		// It may be repeatedly reading after an i/o timeout, for example.
+		// This range will be done only after the receive loop has returned
+		// and closed 'responses' in its defer function.
 		for range responses {
 		}
 
+		// Now cancel all the remaining waiters
 		for _, waiter := range waitingResponses {
 			waiter.ch <- sharedClientResponse{nil, 0, net.ErrClosed}
 			close(waiter.ch)
 		}
+
+		// Drain requests in case they come in while we are closing
+		// down. This loop is done only after 'requests' channel is closed in
+		// SharedClient.close() and it is not possible for new requests or timeouts
+		// to be sent on those closed channels.
```

Comment:
I don't follow this comment. Sending on closed channel panics - surely that's not how we prevent the senders from sending requests? I assume the real check from preventing senders is that `conn` is nil, right?

Reply:
`SharedClient.close()` (note: not a `close(<channel>)`) is called when the shared client can no longer be used for new requests. So there is nothing sent to the closed channel. The point of the comment is that the range loop on the channel completes only after the channel is closed (by the side sending to the channel), so we are guaranteed to send replies to all requests received on this channel.